### DISCUSSION TITLE
Effects with Multiple outputs

### DISCUSSION THREAD

**Discussion by bux225** [pointfreeco/swift-composable-architecture#2451] [Q&A]
I'm doing a standard parse of an API cal for both the status code (for error UX) and for the response.   The below works but I have no idea if it's the right/best way to do this.   I'm having a difficult time writing tests against it. Specifically, the two outputs at the end

In Reducer:

`        case .authenticate:

            return .run { [state = state] send in

                let url = URL(string: "https://account.maxar.com/auth/realms/mds/protocol/openid-connect/token")!
                var request = URLRequest(url: url)
                request.httpMethod = "POST"
                request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
                request.httpBody = "username=\(state.username)&password=\(state.password)&client_id=\(state.clientId)&grant_type=password".data(using: .utf8)
                //print(String(decoding: request.httpBody!, as: UTF8.self))
                let (data, body) = try await URLSession.shared
                    .data(for: request)
                var code = 0
                var response = MGPAuthResponse()
                if let http = body as? HTTPURLResponse {
                    code = http.statusCode
                    if http.statusCode == 200 {
                        response = try JSONDecoder().decode(MGPAuthResponse.self, from: data)
                    }
                }
                await send(.setAuthResponse(response))
                await send(.setHttpStatusCode(code))

            }`

**Comment by mbrandonw** [comment]
Hi @bux225, this code isn't very testable because you are reaching out into the real world to make a network request (the line where you do `URLSession.shared`). Instead you should design a dependency that puts an interface in front of this request and use that instead. Are you familiar with that technique?

**Reply by bux225** [reply]
I appreciate the response.  No, I don't think I'm familiar with that technique.  Would you have a pointer to docs or examples?   Much appreciated

**Reply by mbrandonw** [reply]
We have documentation on dependencies [here](https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/dependencymanagement), and if you follow the [Essentials](https://pointfreeco.github.io/swift-composable-architecture/main/tutorials/meetcomposablearchitecture#essentials) tutorial you will see how to use dependencies from beginning to end.

**Reply by mbrandonw** [reply]
That code is from the ["Side effect"](https://pointfreeco.github.io/swift-composable-architecture/main/tutorials/composablearchitecture/01-02-addingsideeffects) section of the Essentials tutorial. But if you keep reading on, in particular the ["Testing your feature"](https://pointfreeco.github.io/swift-composable-architecture/main/tutorials/composablearchitecture/01-03-testingyourfeature) section, you will see how to control the dependency and make it testable.

**Reply by bux225** [reply]
I see it now in section 4.   Thanks!!

**Comment by bux225** [comment]
OK, I've read and coded the example of numbersapi behind a dependency and the testing.   I still think I have a very naive question but the test doesn't seem to do anything at all.   I can change the url even and the test passes since it's not even testing the call.   Why is this desired?  I'd like to ensure that I could do the following tests on my actual API in my app:
* Test a 200 happy path and that I populate a codable model (tests that the API schema has not changed underneath me)
* Test a 403 to ensure I am providing the user a good UX

**Reply by jshier** [reply]
For the 200 case I suggest you directly test the live version of your networking dependency then, without involving TCA at all. But the 403 case sounds like something you'd simulate and propagate through your stores, so that would be done using typical TCA testing.

You probably want to start a new discussion for this, as it doesn't seem to have anything to do with the topic at hand.

**Reply by mbrandonw** [reply]
Yes, please start a discussion because it is an important topic of why you want to sometimes test with mocks and why you may sometimes want to test against a live API.

**Comment by bux225** [comment]
Will do.   The OP was about multiple effects.  In this case it is the URLResponse status code as well as the data payload itself from an API call.   But I'll reframe into a new discussion
